### PR TITLE
Don't fail when some token prices are missing

### PR DESCRIPTION
**Pull request by mateusz-sekara**
## Motivation

Enforcing all tokens to be present in the job spec creates an implicit dependency between adding tokens to the offramp and creating a job spec with updated token prices. That poses some risk, when the token is added first and then forgotten to be added to job specs or added with some delay, Commit will halt and stop transmitting. Not returning errors in Commit will let us process all the messages. Further, when Exec sees messages with a token without price, it skips those messages and logs an error.

In this case, we favor availability over consistency. Spotting missing tokens in config will be harder, but that won't stop the bridge

## Solution

Replace error with log in PriceGetter

### CODE REVIEW

**Comment by makramkd** [comment]
Two questions:
* Should this behavior be documented in the PriceGetter interface? If we don't error out here I expect us to follow similar behavior in e.g a price getter that uses Mercury/CCIP price serve.
* This is just for price updates right? Merkle roots will still be committed for CCIP sends that involve the token that has a missing price?

File: core/services/ocr2/plugins/ccip/internal/pricegetter/pipeline.go [hunk 1]
```diff
@@ -82,9 +82,11 @@ func (d *PipelineGetter) TokenPricesUSD(ctx context.Context, tokens []common.Add
 
 	// The mapping of token address to source of token price has to live offchain.
 	// Best we can do is sanity check that the token price spec covers all our desired execution token prices.
+	// We don't want to fail here in case of spotting missing token, but rather log it. This should make CCIP more resilient.
+	// In case of missing tokens in the spec, Exec will skip messages containing those tokens.
 	for _, token := range tokens {
 		if _, ok = tokenPrices[token]; !ok {
-			return nil, errors.Errorf("missing token %s from tokensForFeeCoin spec", token)
+			d.lggr.Errorw("missing token from tokensForFeeCoin spec", "token", token.Hex())
```
Comment (makramkd) [inline core/services/ocr2/plugins/ccip/internal/pricegetter/pipeline.go]: In the plugin we have this check: https://github.com/smartcontractkit/ccip/blob/3c25ced34a096e75d1ad948dfc30972406bdc016/core/services/ocr2/plugins/ccip/ccipcommit/ocr2.go#L222-L227.

Won't this error out in the plugin still?

**Comment by mateusz-sekara** [comment]
> Two questions:
> 
> * Should this behavior be documented in the PriceGetter interface? If we don't error out here I expect us to follow similar behavior in e.g a price getter that uses Mercury/CCIP price serve.

Yes, it should be documented, as it's part of the contract. I'm gonna add that

> * This is just for price updates right? Merkle roots will still be committed for CCIP sends that involve the token that has a missing price?

Yes, just price updates. That means that if there is a drift between offramp's tokens and jobspec's tokens, the plugin will keep working and committing. Also, it will keep transmitting prices, but only for tokens that are found.

File: core/services/ocr2/plugins/ccip/internal/pricegetter/pipeline.go (hunk 1 above)
Comment (mateusz-sekara) [reply core/services/ocr2/plugins/ccip/internal/pricegetter/pipeline.go]: Ah you are right. It seems that this check is duplicated - it's both present in the PriceGetter and CommitPlugin. Probably we should get rid of one and keep it only in one place. Maybe the CommitPlugin is the one to decide, not PriceGetter. What do you think @makramkd ?

Comment (makramkd) [reply core/services/ocr2/plugins/ccip/internal/pricegetter/pipeline.go]: I think it makes more sense for the plugin itself to decide since the logic being there is more obvious and usually the first place to check, rather than hidden in the price getter, which feels non-intuitive.

However I think the price getter behavior has to change either way because right now as you noted it'll return `nil` if a single token is missing from the spec.

So maybe gather all the missing tokens in a single `error` and check `errors.Cause` or similar in the plugin?

One idea:

```golang
var ErrMissingToken = errors.New("missing token from tokensForFeeCoin spec")

// The mapping of token address to source of token price has to live offchain.
// Best we can do is sanity check that the token price spec covers all our desired execution token prices.
var errs error
for _, token := range tokens {
	if _, ok = tokenPrices[token]; !ok {
		errs = multierr.Append(errs, errors.Wrap(ErrMissingToken, token.Hex()))
	}
}

return tokenPrices, errs
```

Comment (makramkd) [reply core/services/ocr2/plugins/ccip/internal/pricegetter/pipeline.go]: On the plugin side, maybe you can use `multierr.Every`:

```golang
rawTokenPricesUSD, err := r.priceGetter.TokenPricesUSD(ctx, queryTokens)
if !multierr.Every(err, pricegetter.ErrMissingToken) {
	return nil, nil, err
}
// at this point, if there is an error, its ErrMissingToken
// which is not fatal
lggr.Infow("Raw token prices", "rawTokenPrices", rawTokenPricesUSD, "err", err)
```

Comment (mateusz-sekara) [reply core/services/ocr2/plugins/ccip/internal/pricegetter/pipeline.go]: Hmm, returning both value and error? Not sure about that approach, is it the Go-way? I haven't seen that anywhere in our codebase. I would say that having dedicated errors and then handling them in the Commit makes code more complex than it is supposed to be. 

I think we either want to **be available** - execute as many tokens as we get from PriceGetter, no matter the reason, or **consistent** - fail immediately. `ErrMissingToken` is very custom to the underlying implementation and might not be a good fit for any other implementation (e.g. mercury, lambda, whatever).

If we want to make this interface more generic in this PR and I would suggest the following:
```go
type PriceGetter interface {
        // Must len([]TokenPriceResult) == len(tokens) 
	TokenPricesUSD(ctx context.Context, tokens []common.Address) ([]TokenPriceResult, error)
}

type TokenPriceResult struct {
	TokenAddress common.Address
	Price        *big.Int
	Error error
}
```

This is aligned with the Batching interface. It gives the client control over the returned data and visibility of which tokens failed to fetch. For instance plugin might want to fail when feetoken is missing, but continue when other ones are missing.

Comment (makramkd) [reply core/services/ocr2/plugins/ccip/internal/pricegetter/pipeline.go]: Yeah, that looks good. Top-level error is fatal and each result may error. What I suggested was based off of not changing the interface but I think changing the interface itself in this case is justified.

Comment (makramkd) [reply core/services/ocr2/plugins/ccip/internal/pricegetter/pipeline.go]: You can even add tests for all implementations of `PriceGetter` that assert: `if err == nil { assert.Equal(t, len(tokens), len(tokenPriceResults))` to keep all implementations honest.

**Review by makramkd — APPROVED** [review]
LGTM. Lets also make sure this is called out in the release notes for the next release so that everyone is aware 👍

**Review by RensR — CHANGES_REQUESTED** [review]
Requesting changes to make sure this isn't merged before more discussion. Maybe I missed where we decided this but I'm not fully convinced yet. 

Not starting on missing tokens was an intentional choice, as missing token prices are hard to spot otherwise. While prices are missing, sending the tokens that don't have prices will result in reverting txs on the sending side. 

We lose the invariant that if commit runs, we have all the prices, which I guess will be lost when we move to a dedicated pricing job as well.

File: core/services/ocr2/plugins/ccip/ccipcommit/ocr2.go [hunk 2]
```diff
@@ -236,9 +229,21 @@ func (r *CommitReportingPlugin) generatePriceUpdates(
 		return nil, nil, fmt.Errorf("get tokens decimals: %w", err)
 	}
 
+	// We don't want to fail in case of spotting missing token, but rather log it. This should make CCIP more resilient.
```
Comment (connorwstein) [inline core/services/ocr2/plugins/ccip/ccipcommit/ocr2.go:232]: this is insufficient - there are checks in Report as well (see validateObservations) ensuring we have all the relevant prices. You'd have to remove that `len(destTokens) != len(obs.TokenPricesUSD) ` check and then ensure that calculatePriceUpdates only includes medians of token prices if there is >= f+1 observations of that token price. Has a small drawback that now the f value is scattered. Worth writing a test that covers the observe->report interaction I think.

**Review by connorwstein — CHANGES_REQUESTED** [review]
